packages/docs-site/docs/ref/style/usage.md

```diff
@@ -363,6 +375,85 @@ where `shape_1` and `shape_2` can be variables assigned to shapes.
+- `<FOREACH>` is a semicolon-separated list of object declarations. Objects in the `<FOREACH>` clause are also not collected, but they arrange the `<COLLECT>` objects into groups. The entire `Collector` block runs once for each distinct realization of the `<FOREACH>` clause. The `<FOREACH>` list can contain multiple declarations.
+
+For example, suppose we have the Substance program that defines one set `s1` that contains elements `e1, e2`, and another set `s2` that contains elements `e3, e4, e5`, as follows:
+
+```
+Set s1
+Element e1, e2
+In(e1, s1)
+In(e2, s1)
+
+Set s2
+Element e3, e4, e5
+In(e3, s2)
+In(e4, s2)
+In(e5, s2)
+```
+
+We can write a Collector Block:
+
+```
+collect Element e into es
+where In(e, s)
+foreach Set s {
+    ...
+}
+```
+
+Under the above Substance program, this Collector Block would run twice:
+
+- once with mapping `es -> [e1, e2]` and `s -> s1`
+- once with mapping `es -> [e3, e4, e5]` and `s -> s2`.
+
+Notice that we have splitted the set of all elements `[e1, e2, e3, e4, e5]` into two groups based on the `Set` object that they are contained within.
+
+Within the body of the Collector Blocks, we can do everything that can be done in Selector Blocks, with the one exception that we can only access Style variables in `<INTO>` and `<FOREACH>`.
+
+### Collection Access Expression
+
+Within the Collector Block, the name in the `<INTO>` clause conceptually means a list of Substance objects. We can access the fields of each element in the collection using the "Collection Access" expression with syntax:
+
+```
+listof <FIELD NAME> from <COLLECTION NAME>
+```
+
+This expression takes the `<FIELD NAME>` field of each Substance variable in the collection `<COLLECTION NAME>`, and compiles these fields into an appropriate list.
+
+For example, suppose `elements` is the name in the `<INTO>` block, conceptually representing some list of Substance objects `[e1, e2, ..., en]`. Then, we would expect the expression `listof field from elements` to produce a list `[e1.field, e2.field, ..., en.field]`.
+
+Due to technical constraints, we cannot access _any_ field of each element of the collection. In the above example, if `e1.field` gives a color, then we cannot really put the color into a list, since Penrose does not support lists of colors. We restrict allowable types to as follows:
+
+| Type of `field`    | Collects into                 |
+| ------------------ | ----------------------------- |
+| `FloatV` (number)  | `VectorV` (vector)            |
+| `VectorV` (vector) | `MatrixV` (matrix)            |
+| `ListV` (list)     | `LListV` (list of lists)      |
+| `TupV` (2-tuple)   | `PtListV` (list of 2d points) |
+| some shape         | `ShapeListV` (list of shapes) |
+
+If, in the above example, `e1.field`, `e2.field`, etc. are numbers, then the expresion `listof field from elements` gives us a vector that contains all the values. We can directly plug the vector into accumulation functions such as `average` and `sum`.
```

Comment:
`average` does exist: 
```
  /**
   * Return the average of the floats in the list `xs`.
   */
  average: {
    name: "average",
    description: "Return the average of the floats in the list `xs`.",
    params: [{ name: "xs", type: realNT(), description: "`xs`" }],
    body: (_context: Context, xs: ad.Num[]): FloatV<ad.Num> => {
      return {
        tag: "FloatV",
        contents: div(addN(xs), max(1, xs.length)),
        // To avoid divide-by-0
      };
    },
    returns: valueT("Real"),
  },
```